pkg/storage/redis.go

```diff
@@ -314,15 +313,21 @@ func readAllKeys(ctx context.Context, namespace string, b *RedisDB, pageSize int
 	if pageSize != -1 {
 		scanCount = min(RedisScanBatchSize, pageSize)
 	}
-	for pageSize == -1 || (len(allKeys) < pageSize) {
+	// Scan keys starting at cursor until the end or until we have enough keys
+	for {
 		keys, nextCursor, err = b.db.Scan(ctx, cursor, namespace+"*", int64(scanCount)).Result()
 		if err != nil {
 			return nil, "", errors.Wrap(err, "scan error")
 		}
-
-		allKeys = append(allKeys, keys...)
-
-		if nextCursor == 0 {
+		// Append keys one by one to ensure we don't exceed the page size
+		for _, key := range keys {
```

Comment:
Could this be simplified by using `b.db.Scan(ctx, cursor, namespace+"*", int64(scanCount)).Iterator()` above? 

(as it happens, I also think there might be an implementation error right now, where only a subset of keys are returned)

Reply:
> it does not provide a way to retrieve the current cursor value.

Ah yes, excellent point. 

Changing the function's signature seems like the best approach to me. I suggest keeping the `cursor`, and adding `offsetFromCursor int`. It's important to keep the cursor because it's propagated from the API. 
You'll likely also need to change the way in which the token is parsed in https://github.com/TBD54566975/ssi-service/blob/c8541a8597ab1cd8ad1d650d7d682f58e5119ad5/pkg/storage/redis.go#L37-L43, and how it's encoded in https://github.com/TBD54566975/ssi-service/blob/c8541a8597ab1cd8ad1d650d7d682f58e5119ad5/pkg/storage/redis.go#L332-L335C1

Reply:
> Hence, using Iterator() and keeping the cursor value updated it's not a possibility

I agree. I don't think I was clear with my suggestion, my apologies. What I'm proposing is to keep cursor as a parameter, as well as adding an additional parameter called `offsetFromCursor`. `Scan` would still be used, instead of `Iterator`. Then, the function can start populating the result to return by making a Scan call from the given cursor, and only adding elements if they're greater than or equal to the  `offsetFromCursor`. Does that make sense?

Reply:
The parameter `offsetFromCursor` is something that changes with every request, so I would expect it to be a parameter to the `readAllKeys` function. Can you elaborate on the advantages of putting in in the RedisDB struct? Keep in mind that the RedisDB struct is long lived.

What I was imagining is something were the signature of the function becomes the following:

```go
func readAllKeys(ctx context.Context, namespace string, b *RedisDB, pageSize int, cursor uint64, offsetFromCursor int) (allKeys []string, nextCursor uint64, nextOffsetFromCursor int, err error) {
```

The implementation should add up to `pageSize` elements to `allKeys` by:
1. calling `Scan` with the given `cursor` value
2. adding elements  from the above result only after the `offsetFromCursor` index.

When there are enough elements returned from `Scan` to be able to fill up `allKeys` to the desired `pageSize`, then `nextOffsetFromCursor` should increment by `pageSize`. Otherwise, you should keep iterating over the results from the DB until the previous condition happens.

As far as passing in the `offsetFromCursor` parameter from the callers, below is an example of what I was thinking: 

```go
func (b *RedisDB) ReadPage(ctx context.Context, namespace string, pageToken string, pageSize int) (map[string][]byte, string, error) {
	token := new(PageToken)
	if pageToken != "" {
		var err error
		token, err = parseToken(pageToken)
		if err != nil {
			return nil, "", err
		}
	}

	keys, nextCursor, offsetFromCursor, err := readAllKeys(ctx, namespace, b, pageSize, token.Cursor, token.OffsetFromCursor)
	if err != nil {
		return nil, "", err
	}
	results, err := readAll(ctx, namespace, keys, b)
	if err != nil {
		return nil, "", err
	}
	nextPageToken := PageToken{
		Cursor:           nextCursor,
		OffsetFromCursor: offsetFromCursor,
	}
	encodedToken, err := encodeToken(nextPageToken)
	if err != nil {
		return nil, "", err
	}
	return results, encodedToken, nil
}

type PageToken struct {
	Cursor           uint64
	OffsetFromCursor int
}

func parseToken(pageToken string) (*PageToken, error) {
	pageTokenData, err := base64.RawURLEncoding.DecodeString(pageToken)
	if err != nil {
		return nil, errors.Wrap(err, "decoding page token")
	}

	var token PageToken
	if err := json.Unmarshal(pageTokenData, &token); err != nil {
		return nil, errors.Wrap(err, "unmarshalling page token data")
	}

	return &token, nil
}

func encodeToken(token PageToken) (string, error) {
	data, err := json.Marshal(token)
	if err != nil {
		return "", errors.Wrap(err, "marshalling page token")
	}
	return base64.RawURLEncoding.EncodeToString(data), nil
}
```

Does that all make sense?

Reply:
I think so, yes! This is awesome, thanks so much @radurobot 

Just a note: I don't trust my eyes as much as I would trust tests :)